
# Pandas</p>

*pandas* is a Python library for data analysis. It offers a number of data exploration, cleaning and transformation operations that are critical in working with data in Python.

*pandas* builds upon *numpy* and *scipy* providing easy-to-use data structures and data manipulation functions with integrated indexing.

The main data structures *pandas* provides are *Series* and *DataFrames*. After a brief introduction to these two data structures and data ingestion, the key features of *pandas* that the next three notebooks will cover are:
* Generating descriptive statistics on data
* Data cleaning using built in pandas functions
* Frequent data operations for subsetting, filtering, insertion, deletion and aggregation of data
* Merging multiple datasets using dataframes
* Working with timestamps and time-series data

**Additional Recommended Resources:**
* *pandas* Documentation: http://pandas.pydata.org/pandas-docs/stable/
* *Python for Data Analysis* by Wes McKinney
* *Python Data Science Handbook* by Jake VanderPlas

Let's get started with our first *pandas* notebook!

## Import Libraries
</p>

In [1]:
import pandas as pd


# Introduction to pandas Data Structures</p>

*pandas* has two main data structures it uses, namely, *Series* and *DataFrames*.

## pandas Series</p>

A *pandas Series* is a one-dimensional labeled array. You can also think of it like an ordered dictionary.


In [2]:
ser = pd.Series(data=[100, 200, 300, 400, 500], index=['tom', 'bob', 'nancy', 'dan', 'eric'])

In [3]:
ser

tom      100
bob      200
nancy    300
dan      400
eric     500
dtype: int64

In [4]:
ser.index

Index(['tom', 'bob', 'nancy', 'dan', 'eric'], dtype='object')

In [5]:
ser.loc['nancy']
#Or just ser['nancy'] works as well.

np.int64(300)

In [6]:
ser.iloc[2]

np.int64(300)

In [7]:
'bob' in ser

True

In [8]:
ser * 2

tom       200
bob       400
nancy     600
dan       800
eric     1000
dtype: int64

## Exercise 1:
Suppose that student scores on a homework assignment are give as follows: Alice: 90, Bob: 70, Frank: 20, Mary: 80, and Nicole: 70. Create a series named studentser that has the student name as the index, and the score as the data.

In [9]:
### YOUR CODE HERE! (one line)
studentser = pd.Series(data=[90,70,20,80,70],index=["Alice", "Bob", "Frank", "Mary", "Nicole"])
print (studentser)

Alice     90
Bob       70
Frank     20
Mary      80
Nicole    70
dtype: int64



## pandas DataFrame</p>

*pandas DataFrame* is a 2-dimensional labeled data structure.


### Create DataFrame from dictionary of Python Series</p>

In [10]:
d = {'one' : pd.Series([100., 200., 300.], index=['apple', 'ball', 'clock']),
     'two' : pd.Series([111., 222., 333., 4444.], index=['apple', 'ball', 'cat', 'dog'])}

In [11]:
df = pd.DataFrame(d)
print(df)

         one     two
apple  100.0   111.0
ball   200.0   222.0
cat      NaN   333.0
clock  300.0     NaN
dog      NaN  4444.0


In [12]:
df.index

Index(['apple', 'ball', 'cat', 'clock', 'dog'], dtype='object')

In [13]:
df.columns

Index(['one', 'two'], dtype='object')

### Select a column with either of the following two syntaxes:

In [14]:
# This is called bracket notation and always works.
df['one']

apple    100.0
ball     200.0
cat        NaN
clock    300.0
dog        NaN
Name: one, dtype: float64

In [15]:
# This is called dot notation. It is simpler to read, but will sometimes result in errors
# (if the column name contains a space for example). To be safest, use bracket notation.
df.one

apple    100.0
ball     200.0
cat        NaN
clock    300.0
dog        NaN
Name: one, dtype: float64

### Select a row with either of the following two syntaxes:

In [16]:
df.loc['apple']

one    100.0
two    111.0
Name: apple, dtype: float64

In [17]:
df.iloc[0]

one    100.0
two    111.0
Name: apple, dtype: float64

### Select a specific entry using two indices:

In [18]:
df.iloc[0,0]

np.float64(100.0)

### Get the row number of a specific index using .get_loc

In [19]:
df.index.get_loc('apple')

0

## Exercise 2:
Suppose that in additon to the homework assignment mentioned in Exercise 1, there are also three other homework assignments with the following scores:<br>
Homework 2: Alice: 95, Bob: 100, Frank: 10, Mary: 85, and Nicole: 75<br>
Homework 3: Alice: 90, Bob: 60, Frank: 30, Mary: 80, and Nicole: 75<br>
Homework 4: Alice: 95, Bob: 80, Frank: 40, Mary: 80, and Nicole: 75<br>
Create a dataframe named studentdf contaning the scores for all students on all four homework assignments. Use the student name as the index, and name the columns 'hw 1', 'hw 2', 'hw 3', and 'hw 4'. As above, first create a dictionary of series, and then convert this dictionary to a dataframe.

In [20]:
### YOUR CODE HERE! (two lines)

df2 = {'hw1': pd.Series([90., 70., 20., 80., 70.], index=['Alice', 'Bob', 'Frank', 'Mary', 'Nicole']),
     'hw2' : pd.Series([95., 100., 10., 85., 75.], index=['Alice', 'Bob', 'Frank', 'Mary', 'Nicole']),
     'hw3' : pd.Series([90., 60., 30., 80., 75.], index=['Alice', 'Bob', 'Frank', 'Mary', 'Nicole']),
     'hw4' : pd.Series([95., 80., 40., 80., 75.], index=['Alice', 'Bob', 'Frank', 'Mary', 'Nicole'])
     }
studentdf = pd.DataFrame(df2)

print(studentdf)

         hw1    hw2   hw3   hw4
Alice   90.0   95.0  90.0  95.0
Bob     70.0  100.0  60.0  80.0
Frank   20.0   10.0  30.0  40.0
Mary    80.0   85.0  80.0  80.0
Nicole  70.0   75.0  75.0  75.0



### Basic DataFrame operations</p>

In [21]:
df

one     two
apple  100.0   111.0
ball   200.0   222.0
cat      NaN   333.0
clock  300.0     NaN
dog      NaN  4444.0

In [22]:
df['three'] = df['one'] * df['two']
df

one     two    three
apple  100.0   111.0  11100.0
ball   200.0   222.0  44400.0
cat      NaN   333.0      NaN
clock  300.0     NaN      NaN
dog      NaN  4444.0      NaN

In [23]:
df['flag'] = df['one'] > 250
df

one     two    three   flag
apple  100.0   111.0  11100.0  False
ball   200.0   222.0  44400.0  False
cat      NaN   333.0      NaN  False
clock  300.0     NaN      NaN   True
dog      NaN  4444.0      NaN  False

### We can use Boolean filters with data frames!

In [24]:
df[df['two']>200]
#Note that df.loc[df['two']>200] also works.

one     two    three   flag
ball  200.0   222.0  44400.0  False
cat     NaN   333.0      NaN  False
dog     NaN  4444.0      NaN  False

### Removing a Column

In [25]:
three = df.pop('three')

In [26]:
three

apple    11100.0
ball     44400.0
cat          NaN
clock        NaN
dog          NaN
Name: three, dtype: float64

In [27]:
df

one     two   flag
apple  100.0   111.0  False
ball   200.0   222.0  False
cat      NaN   333.0  False
clock  300.0     NaN   True
dog      NaN  4444.0  False

In [28]:
del df['two']

In [29]:
df

one   flag
apple  100.0  False
ball   200.0  False
cat      NaN  False
clock  300.0   True
dog      NaN  False

### Inserting a Column

In [30]:
df.insert(2, 'copy_of_one', df['one'])
df

one   flag  copy_of_one
apple  100.0  False        100.0
ball   200.0  False        200.0
cat      NaN  False          NaN
clock  300.0   True        300.0
dog      NaN  False          NaN

### Removing a row

In [31]:
df.drop('apple')

one   flag  copy_of_one
ball   200.0  False        200.0
cat      NaN  False          NaN
clock  300.0   True        300.0
dog      NaN  False          NaN

### .sum() and .mean(). axis=0 and axis=1
.sum(axis=0 or 1) will sum elements of a dataframe, and .mean(axis=0 or 1) will compute the mean. axis=0 specifies to compute the sum or mean **of columns** while axis=1 specifies to compute the sum or mean **of rows**.

In [32]:
#Let's make a fresh dataframe:
d = {'one' : pd.Series([100., 200., 300., 400., 500.], index=['apple', 'ball', 'clock', 'cat', 'dog']),
     'two' : pd.Series([111., 222., 333., 4444., 83.], index=['apple', 'ball','clock', 'cat', 'dog'])}
df = pd.DataFrame(d)
print(df)

         one     two
apple  100.0   111.0
ball   200.0   222.0
clock  300.0   333.0
cat    400.0  4444.0
dog    500.0    83.0


In [33]:
#Compute the sum of the two columns and add it to the dataframe in a column called 'total'.
df['total']=df.sum(axis=1)
df

one     two   total
apple  100.0   111.0   211.0
ball   200.0   222.0   422.0
clock  300.0   333.0   633.0
cat    400.0  4444.0  4844.0
dog    500.0    83.0   583.0

In [34]:
#Compute the mean of the two columns and add it to the dataframe in a column called 'average'.
#Note how we select only the columns that we want, avoiding the new 'total' column
df['average']=df.iloc[:,0:2].mean(axis=1)
df

one     two   total  average
apple  100.0   111.0   211.0    105.5
ball   200.0   222.0   422.0    211.0
clock  300.0   333.0   633.0    316.5
cat    400.0  4444.0  4844.0   2422.0
dog    500.0    83.0   583.0    291.5

In [35]:
#Now we can also compute the overall average for all five items:
df['average'].mean(axis=0)

np.float64(669.3)

## Exercise 3:
Add a 'total' column to the studentdf dataframe from Exercise 2 that sums the scores on the four homework assignments. For a check, Nicole's total should be 295.

In [36]:
### YOUR CODE HERE! (one line)
# df['total']=df.sum(axis=1)
studentdf['total']= studentdf.sum(axis=1)
print(studentdf)

         hw1    hw2   hw3   hw4  total
Alice   90.0   95.0  90.0  95.0  370.0
Bob     70.0  100.0  60.0  80.0  310.0
Frank   20.0   10.0  30.0  40.0  100.0
Mary    80.0   85.0  80.0  80.0  325.0
Nicole  70.0   75.0  75.0  75.0  295.0


## Exercise 4
Now add an 'average' column to the studentdf dataframe from Exercise 3 that average the scores on the four homework assignments for each student. See the example above for how to exclude the total column from the average. For a check, Nicole's average should be 73.75.

In [37]:
### YOUR CODE HERE! (one line)
studentdf['average']=studentdf.iloc[:,0:4].mean(axis=1)
print(studentdf)

         hw1    hw2   hw3   hw4  total  average
Alice   90.0   95.0  90.0  95.0  370.0    92.50
Bob     70.0  100.0  60.0  80.0  310.0    77.50
Frank   20.0   10.0  30.0  40.0  100.0    25.00
Mary    80.0   85.0  80.0  80.0  325.0    81.25
Nicole  70.0   75.0  75.0  75.0  295.0    73.75


## Exercise 5
Use a Boolean filter on the studentdf dataframe from Exercise 4 to make a new dataframe called passingdf that contains only those students that are passing (average equal to or greater than 60.0). For a check, this should only remove Frank.

In [38]:
### YOUR CODE HERE! (one line)
passingdf = studentdf[studentdf['average'] >= 60.0]
print(passingdf)

         hw1    hw2   hw3   hw4  total  average
Alice   90.0   95.0  90.0  95.0  370.0    92.50
Bob     70.0  100.0  60.0  80.0  310.0    77.50
Mary    80.0   85.0  80.0  80.0  325.0    81.25
Nicole  70.0   75.0  75.0  75.0  295.0    73.75


## Exercise 6
Compute the overall average of all students that are passing (the average of the four averages above), and store this in a variable called passing_average. The correct passing_average should be 81.25. Make sure to use dataframe operations to compute this (don't just add the 4 numbers and divide by 4).

In [39]:
### YOUR CODE HERE! (one line)
passing_average = passingdf['average'].mean(axis=0)
print(passing_average)

81.25
